crates/re_viewer/src/store_hub.rs

```diff
@@ -197,34 +197,73 @@ impl StoreHub {
     /// Note that the recording is not automatically made active. Use [`StoreHub::set_recording_id`]
     /// if needed.
     pub fn insert_recording(&mut self, store_db: StoreDb) {
-        self.store_dbs.insert_recording(store_db);
+        self.store_bundle.insert_recording(store_db);
     }
 
     /// Mutable access to a [`StoreDb`] by id
     pub fn store_db_mut(&mut self, store_id: &StoreId) -> &mut StoreDb {
-        self.store_dbs.store_db_entry(store_id)
+        self.store_bundle.store_db_entry(store_id)
     }
 
     /// Remove any empty [`StoreDb`]s from the hub
     pub fn purge_empty(&mut self) {
-        self.store_dbs.purge_empty();
+        self.store_bundle.purge_empty();
     }
 
     /// Call [`StoreDb::purge_fraction_of_ram`] on every recording
+    //
+    // NOTE: If you touch any of this, make sure to play around with our GC stress test scripts
+    // available under `$WORKSPACE_ROOT/tests/python/gc_stress`.
     pub fn purge_fraction_of_ram(&mut self, fraction_to_purge: f32) {
-        self.store_dbs.purge_fraction_of_ram(fraction_to_purge);
+        re_tracing::profile_function!();
+
+        let Some(store_id) = self.store_bundle.find_oldest_modified_recording().cloned() else {
+            return;
+        };
+
+        let store_dbs = &mut self.store_bundle.store_dbs;
+        if store_dbs.len() <= 1 {
```

Comment:
Surprised to see this return early when there is 1 store_db? Is this guaranteed to be some kind of special store that we don't want to GC? Please clarify with a comment if so.

Reply:
Uh-oh. No that's just me having shuffled things around one time too many :grimacing: Nice catch.

Reply:
Hmmm, I went with the following... opinions welcome.

```diff
commit 8e6b8853b3751037989e80c26704fbf3d0438a64
Author: Clement Rey <cr.rey.clement@gmail.com>
Date:   Wed Nov 8 19:21:51 2023 +0100

    always GC, but dont remove the last one

diff --git a/crates/re_viewer/src/store_hub.rs b/crates/re_viewer/src/store_hub.rs
index e17fd0d0dd..3457abc1b6 100644
--- a/crates/re_viewer/src/store_hub.rs
+++ b/crates/re_viewer/src/store_hub.rs
@@ -222,9 +222,6 @@ impl StoreHub {
         };
 
         let store_dbs = &mut self.store_bundle.store_dbs;
-        if store_dbs.len() <= 1 {
-            return;
-        }
 
         let Some(store_db) = store_dbs.get_mut(&store_id) else {
             if cfg!(debug_assertions) {
@@ -239,9 +236,21 @@ impl StoreHub {
         let store_size_after =
             store_db.store().timeless_size_bytes() + store_db.store().temporal_size_bytes();
 
+        // No point keeping an empty recording around.
+        if store_db.is_empty() {
+            self.remove_recording_id(&store_id);
+            return;
+        }
+
         // Running the GC didn't do anything.
-        // That's because all that's left in that store is protected rows: it's time to remove it entirely.
-        if store_size_before == store_size_after {
+        //
+        // That's because all that's left in that store is protected rows: it's time to remove it
+        // entirely, unless it's the last recording still standing, in which case we're better off
+        // keeping some data around to show the user rather than a blank screen.
+        //
+        // If the user needs the memory for something else, they will get it back as soon as they
+        // log new things anyhow.
+        if store_size_before == store_size_after && store_dbs.len() > 1 {
             self.remove_recording_id(&store_id);
         }
 
```